src/gnatss/configs/main.py

```diff
@@ -5,18 +5,72 @@
 """
 import warnings
 from pathlib import Path
-from typing import Any, Dict, Optional
+from typing import Any, Dict, Optional, Type
 
 import yaml
-from pydantic import BaseSettings, Field
-from pydantic.fields import ModelField
+from pydantic import Field
+from pydantic.fields import FieldInfo
+from pydantic_settings import (
+    BaseSettings,
+    PydanticBaseSettingsSource,
+    SettingsConfigDict,
+)
 
 from .io import OutputPath
 from .solver import Solver
 
 CONFIG_FILE = "config.yaml"
 
 
+class YamlConfigSettingsSource(PydanticBaseSettingsSource):
```

Comment:
Can we add docstrings for functions defined in this class? I am not sure about what the function does. Please correct me if I am wrong and these are some generic function categories defined when we use Pydantic.